### ISSUE TITLE
Feature request: stateless API

### ISSUE THREAD

**Issue by jonasN5** [pennersr/django-allauth#3805]
The current implementation heavily relies on sessions. I'd like to make the usage of sessions optional, which is often the case for mobile apps that can use a stateless API.
Thoughts?

**Comment by pennersr** [comment]
The authentication process is inherently stateful, I do not see how you can remove that state out of the equation. For example:
- When a user signed in but has yet to complete 2FA -- that is state that needs to be recorded somewhere.
- Whether or not the re-authentication is required depends on when the user last authenticated -- which is again state.
- When signing in using an OTP that is mailed to the user, we need to keep track of that.

So state is is there, and there to stay. What could be discussed is where you store that state. For example, you could keep signed payloads over at the client, but that is effectively what Django's signed cookies approach for sessions is doing. So, an interesting approach would be to switch your session backend over to that and make that work.

> I'd like to make the usage of sessions optional

Sessions are required during the authentication phase, they cannot be dropped, you can only discuss where they are stored. Once authenticated, you can go fully stateless by issuing e.g. a JWT access token and use that throughout your application.

**Comment by jonasN5** [comment]
My bad let me rephrase: I'd like to make the usage of sessions optional for
the registration phase. For instance, when using jets and no 2FA, just a
sign up with apple, there is no need for a session. A code is provided and
in exchange an access token is returned to the mobile app. Yes, there are
cases that require sessions or some form of server side storing mechanism
but also cases (simpler ones) that do not. For those simple cases, sessions
are a bunch of overhead, both in CPU and database calls.

**Comment by jonasN5** [comment]
Jwts*, not jets -_-

**Comment by pennersr** [comment]
> both in CPU and database calls.

Very few people are at a scale where this is an issue. And, database calls can be completely avoiding by using Redis/Redict based sessions.

But you are right, depending on the requirements, you may be able to strip out certain allauth functionality and reach a minimal core that can be completely session free, but if that is what you are after then I don't think allauth is meant for your use case.  Supporting that in allauth would break a lot of regular functionality, without any tangible benefit.

**Comment by Aniket-Singla** [comment]
I vote for this use case as well. One option that I can suggest is to try moving the piece of code that deals with setting and getting values from session to Adapters where the default adapter relies on session.

**Comment by pennersr** [comment]
@Aniket-Singla Django's sessions are already pluggable, so I don't really see how redoing that in adapter code would help.

**Comment by pennersr** [comment]
https://codeberg.org/allauth/django-allauth/issues/3805
